diff -Naur orginal/include/KNI/cdlCOM.h modified/include/KNI/cdlCOM.h
--- orginal/include/KNI/cdlCOM.h	2007-04-12 15:52:07.000000000 +0200
+++ modified/include/KNI/cdlCOM.h	2009-11-19 13:45:04.000000000 +0100
@@ -107,7 +107,7 @@
 	 *	describes the desired serial port. An attempt to open a connection
 	 *	to the desired device will be tried.
 	 */
-	CCdlCOM(TCdlCOMDesc ccd);
+	CCdlCOM(TCdlCOMDesc ccd, const std::string& deviceName = "");
 
 	/*! \brief	Destructs the class
 	 */	virtual ~CCdlCOM();
diff -Naur orginal/src/Base/cdlCOM.cpp modified/src/Base/cdlCOM.cpp
--- orginal/src/Base/cdlCOM.cpp	2007-04-12 15:52:07.000000000 +0200
+++ modified/src/Base/cdlCOM.cpp	2009-11-19 13:45:19.000000000 +0100
@@ -159,21 +159,21 @@
 
 #else //LINUX
 
-CCdlCOM::CCdlCOM(TCdlCOMDesc ccd) : _deviceName(""), _ccd(), _prtHdl(-1), _oto() {
+CCdlCOM::CCdlCOM(TCdlCOMDesc ccd, const std::string& deviceName) : _deviceName(deviceName), _ccd(), _prtHdl(-1), _oto() {
 
 	int prtHdl	= -1;
 
-	std::string deviceName;
+	if (_deviceName.length() == 0)
+	{
+		_deviceName = "/dev/ttyS" + digit(ccd.port);
+	}
 	struct termios	nto, oto;
-	char		name[11];
-
 	errno = 0;
 
-	strncpy(name, "/dev/ttyS ", 11);
-	name[9] = digit(ccd.port);
+	char*   name = new char[_deviceName.length() +1];
+	strncpy(name, _deviceName.c_str(), _deviceName.length());
 	prtHdl = ::open(name, O_RDWR | O_NOCTTY | O_NDELAY| O_NONBLOCK);
-
-	_deviceName = name;
+	delete name;
 
 	if (prtHdl < 0) {
 		throw CannotOpenPortException(_deviceName, strerror(errno));
@@ -289,7 +289,6 @@
 	}
 
 	_prtHdl = prtHdl;
-	_deviceName = deviceName;
 	_ccd = ccd;
 	_oto = oto;
 }
diff -Naur orginal/src/Base/kmlExt.cpp modified/src/Base/kmlExt.cpp
--- orginal/src/Base/kmlExt.cpp	2009-01-28 11:50:19.000000000 +0100
+++ modified/src/Base/kmlExt.cpp	2010-03-17 13:39:28.000000000 +0100
@@ -103,7 +103,7 @@
 	int gripperOpenEncoders, gripperCloseEncoders;
 	infos->getGripperParameters(gripperIsPresent, gripperOpenEncoders, gripperCloseEncoders);
 	setGripperParameters(gripperIsPresent, gripperOpenEncoders, gripperCloseEncoders);
-	
+
 }
 
 void CKatana::create(TKatGNL& gnl, TKatMOT& mot, TKatSCT& sct, TKatEFF& eff, CCplBase* protocol) {
@@ -146,7 +146,7 @@
 		}
 		while(buf[2] == 4);
 		std::cout << "...done with calibration.\n";
-		
+
 	}
 	else if(mKatanaType == 300){
 		std::cout << "Katana300 calibration started\n";
@@ -316,14 +316,14 @@
 			setForceLimit(i, limit);
 		}
 	}
-	
+
 	// check axis number
 	if (axis < 1) return;
 	if (axis > getNumberOfMotors()) return;
-	
+
 	// can not set current limit on drive controller
 	if (base->GetMOT()->arr[axis-1].GetSFW()->type == 0) return;
-	
+
 	int force = abs(limit);
 	if (force > 100)
 		force = 100;
@@ -493,10 +493,14 @@
 		poll_t.Start();
 		base->recvMPS(); // get position for all motors
 		base->recvGMS(); // get status flags for all motors
-		for (int idx=0; idx<getNumberOfMotors(); idx++) {
+
+//		for (int idx=0; idx<getNumberOfMotors(); idx++) {  // original code
+
+		// only check for those encoders which were set
+		for (int idx=0; idx<i; idx++) {
 			if (mot->arr[idx].GetPVP()->msf == 40)
 				throw MotorCrashException();
-			pos_reached &= std::abs(mot->arr[idx].GetTPS()->tarpos - mot->arr[idx].GetPVP()->pos) < 100;
+			pos_reached &= std::abs(mot->arr[idx].GetTPS()->tarpos - mot->arr[idx].GetPVP()->pos) < encTolerance;
 		}
 		if (pos_reached)
 			break;
@@ -550,14 +554,14 @@
 CKatana::openGripper(bool waitUntilReached, int waitTimeout) {
 	if(!_gripperIsPresent)
 		return;
-	moveMotorToEnc( getNumberOfMotors()-1, _gripperOpenEncoders, waitUntilReached, waitTimeout );
+	moveMotorToEnc( getNumberOfMotors()-1, _gripperOpenEncoders, waitUntilReached, 100, waitTimeout );
 }
 
 void
 CKatana::closeGripper(bool waitUntilReached, int waitTimeout) {
 	if(!_gripperIsPresent)
 		return;
-	moveMotorToEnc( getNumberOfMotors()-1, _gripperCloseEncoders, waitUntilReached, waitTimeout );
+	moveMotorToEnc( getNumberOfMotors()-1, _gripperCloseEncoders, waitUntilReached, 100, waitTimeout );
 }
 
 
diff -Naur orginal/src/Base/Makefile modified/src/Base/Makefile
--- orginal/src/Base/Makefile	2008-11-18 07:01:36.000000000 +0100
+++ modified/src/Base/Makefile	2009-11-12 16:05:25.000000000 +0100
@@ -7,7 +7,7 @@
 CXX		=	g++
 CXXFLAGS	+=	-fPIC -g -O2 -Wall
 
-TARGETS		=	alib solib
+TARGETS		=	alib #solib
 
 LIBNAME = libKNIBase
 SOURCES = cdlCOM.cpp cplSerial.cpp cdlSocket.cpp CRC.cpp kmlBase.cpp kmlExt.cpp kmlMotBase.cpp kmlSctBase.cpp kmlFactories.cpp
diff -Naur orginal/src/InvKin/Makefile modified/src/InvKin/Makefile
--- orginal/src/InvKin/Makefile	2008-11-18 07:01:36.000000000 +0100
+++ modified/src/InvKin/Makefile	2009-11-12 16:05:47.000000000 +0100
@@ -7,7 +7,7 @@
 CXX		=	g++
 CXXFLAGS	+=	-fPIC -g -O2 -Wall
 
-TARGETS		=	alib solib
+TARGETS		=	alib #solib
 
 LIBNAME = libKNI_InvKin
 SOURCES = ikBase.cpp KatanaKinematics6M90G.cpp KatanaKinematics6M90T.cpp KatanaKinematics6M180.cpp KatanaKinematics5M180.cpp KatanaKinematicsDecisionAlgorithms.cpp
diff -Naur orginal/src/kni_wrapper/Makefile modified/src/kni_wrapper/Makefile
--- orginal/src/kni_wrapper/Makefile	2008-11-18 07:01:36.000000000 +0100
+++ modified/src/kni_wrapper/Makefile	2009-11-12 16:05:47.000000000 +0100
@@ -7,7 +7,7 @@
 CXX		=	g++
 CXXFLAGS	+=	-fPIC -g -O2 -Wall
 
-TARGETS		=	alib solib
+TARGETS		=	alib #solib
 
 LIBNAME = kni
 SOURCES = kni_wrapper.cpp
diff -Naur orginal/src/LM/Makefile modified/src/LM/Makefile
--- orginal/src/LM/Makefile	2008-11-18 07:01:36.000000000 +0100
+++ modified/src/LM/Makefile	2009-11-12 16:05:47.000000000 +0100
@@ -7,7 +7,7 @@
 CXX		=	g++
 CXXFLAGS	+=	-fPIC -g -O2 -Wall
 
-TARGETS		=	alib solib
+TARGETS		=	alib #solib
 
 LIBNAME = libKNI_LM
 SOURCES = lmBase.cpp
